### ISSUE TITLE
All Firestore writes prevented by 2x"Missing or insufficient permissions" in quick succession

### ISSUE THREAD

**Issue by Ruberik** [firebase/firebase-ios-sdk#835]
* Xcode version: 9.2
  * Firebase SDK version: 4.9.0 (also appears to be a problem for 4.8.0)
  * Firebase Component: Firestore
  * Component version: 0.10.1 (also appears to be a problem for 0.9.4)

### Steps to reproduce:

When multiple "Missing or insufficient permissions" errors occur in quick succession, the client stops successfully writing to Firestore, and apparently stops trying to.

I'll provide specific code later, but the outline is:
1. Set up rules with a collection that can't be written to, foo, and a collection that can be written to, bar.
2. Write something to "bar" to confirm you can.
3. In a loop, write something to "foo" three times.
4. Confirm you can no longer write to "bar", or even get a callback with an error from attempting to write to "foo".

#### Firestore rules:

```
service cloud.firestore {
  match /databases/{database}/documents {
    match /bar/{id} {
      allow read, write: if true;
    }
    match /foo/{id} {
      allow read, write: if false;
    }
  }
}
```

#### Code:

1. Take quickstart-ios's firestore "FriendlyEats" quickstart.
2. Replace the following two functions in RestaurantsTableViewController as follows:

```
  @IBAction func didTapPopulateButton(_ sender: Any) {
    NSLog("Populate tapped")
    let uid = Auth.auth().currentUser!.uid
    Firestore.firestore().collection("bar").document().setData(["a": 5]) {
      NSLog("bar result: \($0)")
    }
  }

  @IBAction func didTapFilterButton(_ sender: Any) {
    NSLog("Filter button tapped")
    for _ in [1, 2, 3] {
      Firestore.firestore().collection("foo").document().setData(["a": 5]) {
        NSLog("foo result: \($0?.localizedDescription)")
      }
    }
  }
```

3. In AppDelegate, *optionally* add some or all of:
```
    Firestore.enableLogging(true)
    let settings = FirestoreSettings()
    settings.isPersistenceEnabled = false
    Firestore.firestore().settings = settings
```
This bug appears to be independent of whether offline persistence is enabled.

4. Start the app, and log in.

5. Click "populate" and check your database to confirm that something was written in the "bar" collection.

6. Click "filter" and confirm that you get a bunch of "Missing or insufficient permissions" errors.

7. Click "populate" and "filter" and confirm that you don't get any more callbacks or any database writes; just "X button tapped" log messages.

**Comment by mikelehen** [comment]
Thanks very much for the detailed report. We found this issue ourselves a few days ago and the fix was checked in as https://github.com/firebase/firebase-ios-sdk/pull/810.  There was indeed a race condition when rapidly doing writes after a permission denied failure.  The fix should be included in the next Firebase iOS SDK release.  Thanks again and sorry for your trouble!

**Comment by wilhuff** [comment]
(reopening because we haven't formally adopted https://github.com/firebase/firebase-ios-sdk/issues/819 yet)

**Comment by ryanwilson** [comment]
This fix is now live as part of the [Firebase 4.10.0](https://github.com/firebase/firebase-ios-sdk/commits/release-4.10.0) release.
